File: tests/xcalablemp/NOT-IMPLEMENTED/59/reduction_async20.c
/*testp95.c*/
/*loop指示文とreduction指示文のテスト*/
#include<mpi.h>
#include<xmp.h>
#include<stdio.h>   
#pragma xmp nodes p(4)
static const int N=1000;
#pragma xmp template t(0:N-1,0:N-1,0:N-1)
#pragma xmp distribute t(*,*,gblock((/333,555,111,1))) onto p
int a1[N],a2[N],aa;
#pragma xmp align a1[i] with t(*,*,i)
#pragma xmp align a2[i] with t(*,*,i)
int i;
char *result;
int main(void){
#pragma xmp loop on t(:,:,i)
   for(i=0;i<N;i++){
      a1[i] = 0;
      a2[i] = 0;
   }
   result = "OK";
#pragma xmp loop on t(:,:,i)
   for(i=0;i<N;i++){
      aa = a2[i];
#pragma xmp reduction(||: aa) async(1)
      if(a1[i] != 0) result = "NG";
#pragma xmp wait_async(1)
      aa = aa-1;
      a1[i] = aa;
   }
#pragma xmp loop on t(:,:,i)
   for(i=0;i<N;i++)
      if(a1[i] != i-1) result = "NG";

   printf("%d %s %s\n",xmp_node_num(),"testp095.c",result); 
   return 0;
}
      
         
      
   

File: tests/XMP/NOT-IMPLEMENTED/59/reduction_async20.f90
! testp095.f
! loop指示文とreduction指示文のテスト (1次元では意味なし？)

      program main
      include 'xmp_lib.h'
      integer,parameter:: N=1000
!$xmp nodes p(4)
!$xmp template t(N,N,N)
!$xmp distribute t(*,*,gblock((/333,555,111,1/))) onto p
      integer a1(N), a2(N), aa
!$xmp align a1(i) with t(*,*,i)
!$xmp align a2(i) with t(*,*,i)
      character(len=2) result

!$xmp loop on t(:,:,i)
      do i=1, N
         a1(i) = 0
         a2(i) = i
      enddo
      
      result = 'OK'
!$xmp loop on t(:,:,i)
      do i=1, N
         aa = a2(i)
!$xmp reduction(ior: aa) async(1)
         if(a1(i) .ne. 0) result = 'NG'
!$xmp wait_async(1)
         aa = aa-1
         a1(i) = aa
      enddo

!$xmp loop on t(:,:,i)
      do i=1, N
         if(a1(i) .ne. i-1) result = 'NG'
      enddo

      print *, xmp_node_num(), 'testp095.f ', result

      end
      
